src/driver-roccat-kone-aimo.c

```diff
@@ -0,0 +1,1220 @@
+
+
+	rc = ratbag_hidraw_set_feature_report(device, ROCCAT_REPORT_ID_SETTINGS,
+						  (uint8_t*)report, ROCCAT_REPORT_SIZE_SETTINGS);
+
+	if (rc < 0)
+		return rc;
+
+	if (rc != ROCCAT_REPORT_SIZE_SETTINGS)
+		return -EIO;
+
+	rc = roccat_wait_ready(device);
+	if (rc) {
+		log_error(device->ratbag,
+			  "Error while waiting for the device to be ready: %s (%d)\n",
+			  strerror(-rc), rc);
+	}
+
+	rc = ratbag_hidraw_set_feature_report(device, ROCCAT_REPORT_ID_KEY_MAPPING,
+						  (uint8_t*)buttons, ROCCAT_REPORT_SIZE_BUTTONS);
+
+	if (rc < 0)
+		return rc;
+
+	if (rc != ROCCAT_REPORT_SIZE_BUTTONS)
+		return -EIO;
+
+	rc = roccat_wait_ready(device);
+	if (rc) {
+		log_error(device->ratbag,
+			  "Error while waiting for the device to be ready: %s (%d)\n",
+			  strerror(-rc), rc);
+	}
+
+	log_debug(device->ratbag, "profile: %d written %s:%d\n",
+		profile->index,
+		__FILE__, __LINE__);
+
+	return rc;
+}
+
+static void roccat_read_macro(struct roccat_macro* macro, struct ratbag_button* button) {
+	struct ratbag_button_macro *m = NULL;
+	unsigned j, time;
+
+	char name[ROCCAT_MACRO_NAME_LENGTH+1] = { '\0' };
+	strncpy(name, macro->name, ROCCAT_MACRO_NAME_LENGTH);
+
+	m = ratbag_button_macro_new(name);
+	// libratbag does offer API for macro groups
+	m->macro.group = (char*)zalloc(ROCCAT_MACRO_GROUP_NAME_LENGTH+1);
+	strncpy(m->macro.group, macro->group, ROCCAT_MACRO_GROUP_NAME_LENGTH);
+
+	log_debug(button->profile->device->ratbag,
+		"macro on button %d of profile %d is named '%s' (from folder '%s'), and contains %d events:\n",
+		button->index, button->profile->index,
+		name, m->macro.group, macro->length);
+	// libratbag can't keep track of the whole macro (MAX_MACRO_EVENTS)
+	// In libratbag, each event is implemented as two separate (KEY_PRESS/KEY_RELEASE and WAIT)
+	for (j = 0; j < macro->length && j < MAX_MACRO_EVENTS/2; j++) {
+		unsigned int keycode = ratbag_hidraw_get_keycode_from_keyboard_usage(button->profile->device,
+						macro->keys[j].keycode);
+		ratbag_button_macro_set_event(m,
+							j * 2,
+							macro->keys[j].flag & 0x01 ? RATBAG_MACRO_EVENT_KEY_PRESSED : RATBAG_MACRO_EVENT_KEY_RELEASED,
+							keycode);
+		if (macro->keys[j].time)
+			time = macro->keys[j].time;
+		else
+			time = macro->keys[j].flag & 0x01 ? 10 : 50;
+		ratbag_button_macro_set_event(m,
+							j * 2 + 1,
+							RATBAG_MACRO_EVENT_WAIT,
+							time);
+
+		log_debug(button->profile->device->ratbag,
+			"    - %s %s\n",
+			libevdev_event_code_get_name(EV_KEY, keycode),
+			macro->keys[j].flag == 0x02 ? "released" : "pressed");
+	}
+	ratbag_button_copy_macro(button, m);
+	ratbag_button_macro_unref(m);
+}
+
+static void
+roccat_read_button(struct ratbag_button *button)
+{
+	const struct ratbag_button_action *action;
+	struct ratbag_device *device = button->profile->device;
+	struct roccat_data *drv_data = ratbag_get_drv_data(device);
+	struct roccat_macro *macro;
+	int rc;
+
+	action = roccat_button_to_action(button->profile, button->index);
```

Comment:
I think this may have been intended to be:
```suggestion
	button->action = roccat_button_to_action(button->profile, button->index);
```

I think this is the reason libratbag says "UNKNOWN" on buttons.

Reply:
Maybe, the next lines do that. it calls `ratbag_button_set_action` which does a `button->action = *action;` and a `button->action.macro = macro` if its a macro.

Line 1229 in `libratbag.c`

Reply:
Ah, literally the next line. Not thorough review, sorry...
Of course it's better to use the wrapper instead of directly modifying the underlying data.